mailbox/plugin/deleted-messages-vault-cassandra/src/main/java/org/apache/james/vault/metadata/CassandraDeletedMessageMetadataVault.java

```diff
@@ -0,0 +1,84 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault.metadata;
+
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MessageId;
+import org.reactivestreams.Publisher;
+
+public class CassandraDeletedMessageMetadataVault implements DeletedMessageMetadataVault {
+    private final MetadataDAO metadataDAO;
+    private final BucketListDAO bucketListDAO;
+    private final StorageInformationDAO storageInformationDAO;
+    private final UserPerBucketDAO userPerBucketDAO;
+
+    CassandraDeletedMessageMetadataVault(MetadataDAO metadataDAO, BucketListDAO bucketListDAO, StorageInformationDAO storageInformationDAO, UserPerBucketDAO userPerBucketDAO) {
+        this.metadataDAO = metadataDAO;
+        this.bucketListDAO = bucketListDAO;
+        this.storageInformationDAO = storageInformationDAO;
+        this.userPerBucketDAO = userPerBucketDAO;
+    }
+
+    @Override
+    public Publisher<Void> store(DeletedMessageWithStorageInformation deletedMessage) {
+        BucketName bucketName = deletedMessage.getStorageInformation().getBucketName();
+        User owner = deletedMessage.getDeletedMessage().getOwner();
+        MessageId messageId = deletedMessage.getDeletedMessage().getMessageId();
+        return metadataDAO.store(deletedMessage)
+            .then(userPerBucketDAO.addUser(bucketName, owner))
+            .then(storageInformationDAO.referenceStorageInformation(owner, messageId, deletedMessage.getStorageInformation()))
+            .then(bucketListDAO.addBucket(bucketName));
+    }
+
+    @Override
+    public Publisher<Void> removeMetadataRelatedToBucket(BucketName bucketName) {
```

Comment:
If I understand correctly, there's no way to ensure the bucket is deleted in the end? (what if there's a concurrent process adding things into any of these tables?)

Reply:
That's not a problem for the vault feature: we do delete only buckets that we do not use for long.

Reply:
As far as I know it's a property that is enforced nowhere. When we encode such an assumption into the code without having any way to check it in the future, we take the risk of breaking the assumption later without we even notice. It's brittle code IMO.

Reply:
Well anyway the result of concurrent addition / deletion is dependant of operation ordering.